### PR TITLE
Constify multiplication (Mul, MulAssign)

### PR DESCRIPTION
**Pull request by kaidokert**
Multiplication core const #58

## Summary by Sourcery

Const-enable FixedUInt multiplication operations by introducing a reusable const-compatible array multiplication helper and rewiring Mul/MulAssign to use it.

New Features:
- Expose a standalone const-compatible array multiplication function for FixedUInt backing arrays.
- Support const evaluation of FixedUInt Mul and MulAssign implementations across owned and referenced variants.

Enhancements:
- Refactor FixedUInt internal multiplication to delegate to the shared const_array_mul helper while preserving overflow detection semantics.

Tests:
- Add unit tests for the const array multiplication helper, including overflow and carry-propagation scenarios, and a nightly-only const-eval check for multiplication results.

## Summary by CodeRabbit

* **New Features**
  * Multiplication now supports const (compile-time) evaluation across more operand forms, including MulAssign and reference combinations.
  * Added a const-capable, standalone multiplication utility exposed for testing and broader const use.
  * Added public export of an additional bounded trait for improved API compatibility.

* **Behavior Change**
  * Const-time multiplication detects overflow and will panic in const contexts.

* **Tests**
  * Expanded tests for multiplication covering basic, carry, and overflow scenarios (including const/nightly cases).

### CODE REVIEW

File: src/fixeduint.rs [hunk 1]
```diff
@@ -674,6 +641,60 @@ c0nst::c0nst! {
...
         }
     }
+
+    /// Standalone const-compatible array multiplication (no FixedUInt dependency)
+    /// Returns (result_array, overflowed)
+    pub(crate) c0nst fn const_array_mul<T: [c0nst] ConstMachineWord, const N: usize, const CHECK_OVERFLOW: bool>(
+        op1: &[T; N],
+        op2: &[T; N],
+        word_bits: usize,
+    ) -> ([T; N], bool) {
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint.rs]: There's no surrounding impl - this is a free function in a `c0nst`  block

File: src/fixeduint.rs [hunk 2]
```diff
@@ -351,43 +351,10 @@ c0nst::c0nst! {
...
-                    }
-                }
-            }
-            if !carry.is_zero() && CHECK_OVERFLOW {
-                overflowed = true;
-            }
-        }
-        (result, overflowed)
+        let (array, overflowed) =
+            const_array_mul::<T, N, CHECK_OVERFLOW>(&op1.array, &op2.array, Self::WORD_BITS);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint.rs]: const_array_mul IS a free function - inside of `c0nst` block. It is NOT an impl

**Comment by kaidokert** [comment]
@sourcery-ai review

File: src/fixeduint.rs [hunk 3]
```diff
@@ -674,6 +641,56 @@ c0nst::c0nst! {
...
+    pub(crate) c0nst fn const_array_mul<T: [c0nst] ConstMachineWord, const N: usize, const CHECK_OVERFLOW: bool>(
+        op1: &[T; N],
+        op2: &[T; N],
+        word_bits: usize,
+    ) -> ([T; N], bool) {
+        let mut result: [T; N] = [<T as ConstZero>::zero(); N];
+        let mut overflowed = false;
+        let t_max = <T as ConstMachineWord>::to_double(<T as ConstBounded>::max_value());
+        // Zero for double word type
+        let dw_zero: <T as ConstMachineWord>::ConstDoubleWord = core::convert::From::from(0u8);
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint.rs]: done
